pkg/cloud/aws/services/ec2/subnets.go

```diff
@@ -175,14 +179,59 @@ func (s *Service) describeVpcSubnets() (v1alpha1.Subnets, error) {
 		return nil, errors.Wrapf(err, "failed to describe subnets in vpc %q", s.scope.VPC().ID)
 	}
 
+	routeTables, err := s.describeVpcRouteTablesBySubnet()
+	if err != nil {
+		return nil, err
+	}
+
+	natGateways, err := s.describeNatGatewaysBySubnet()
+	if err != nil {
+		return nil, err
+	}
+
 	subnets := make([]*v1alpha1.SubnetSpec, 0, len(out.Subnets))
+	// Besides what the AWS API tells us directly about the subnets, we also want to discover whether the subnet is "public" (i.e. directly connected to the internet) and if there are any associated NAT gateways.
+	// We also look for a tag indicating that a particular subnet should be public, to try and determine whether a managed VPC's subnet should have such a route, but does not.
 	for _, ec2sn := range out.Subnets {
+		var hasPublicTag bool
+		for _, tag := range ec2sn.Tags {
+			if tag == nil {
+				continue
+			}
+
+			if tag.Key != nil && *tag.Key == NameSubnetRoutable && tag.Value != nil && *tag.Value == "public" {
+				hasPublicTag = true
+			}
+		}
+
+		var hasPublicRoute bool
+		var routeTableId *string
+		rt := routeTables[*ec2sn.SubnetId]
+		if rt != nil {
+			routeTableId = rt.RouteTableId
+			for _, route := range rt.Routes {
+				if route.GatewayId == nil {
+					continue
+				}
+				if strings.HasPrefix(*route.GatewayId, "igw") {
+					hasPublicRoute = true
+				}
+			}
+		}
+
+		ngw := natGateways[*ec2sn.SubnetId]
+		var natGatewayId *string
+		if ngw != nil {
+			natGatewayId = ngw.NatGatewayId
+		}
 		subnets = append(subnets, &v1alpha1.SubnetSpec{
```

Comment:
If we move the definition of SubnetSpec on line 196, when the for loop starts, we'll be able to assign directly to the struct's fields and remove the temporary vars.